File: dotnet/DesignPatterns/CreationalDesignPatterns/PrototypeDesignPattern/PrototypeDesignPattern.Example3/Shape.cs
namespace PrototypeDesignPattern.Example3;

public abstract class Shape : ICloneable
{
	public string Color { get; set; } = string.Empty;
	public double X { get; set; }
	public double Y { get; set; }

	protected Shape(string color, double x, double y)
	{
		Color = color;
		X = x;
		Y = y;
	}

	public abstract object Clone();

	public virtual void Display()
	{
		Console.WriteLine($"Shape at ({X}, {Y}) with color {Color}");
	}
}

File: nodejs/DesignPatterns/CreationalDesignPatterns/PrototypeDesignPattern/PrototypeDesignPattern.Example3/src/Shape.ts
export abstract class Shape {
  color: string;
  x: number;
  y: number;

  protected constructor(color: string, x: number, y: number) {
    this.color = color;
    this.x = x;
    this.y = y;
  }

  abstract clone(): Shape;

  display(): void {
    console.log(`Shape at (${this.x}, ${this.y}) with color ${this.color}`);
  }
}
